src/na/na_ofi.c

```diff
@@ -2600,13 +2703,37 @@ na_ofi_cq_read(na_context_t *context, size_t max_count,
 //                         cq_err.flags);
             goto out;
 
-        case FI_EADDRNOTAVAIL:
+        case FI_EADDRNOTAVAIL: {
+            struct na_ofi_class *priv = NA_OFI_CLASS(na_class);
+            struct fid_av *av_hdl = priv->domain->fi_av;
+            void *err_addr = NULL;
+            size_t err_addrlen;
+
+            /* Copy addr information */
+            err_addr = malloc(cq_err.err_data_size);
+            NA_CHECK_ERROR(err_addr == NULL, out, ret, NA_NOMEM_ERROR,
+                "Could not allocate err_addr");
+            err_addrlen = cq_err.err_data_size;
+            memcpy(err_addr, cq_err.err_data, err_addrlen);
+
```

Comment:
I don't think that block of code is needed anymore, I removed it previously so that we don't need the extra malloc. The av insert is done in the unexpected event callback.

Reply:
without this, the na_ofi_cq_process_recv_unexpected_event() function will enter the branch: ```else if (src_err_addr && src_err_addrlen)```, and in there ```src_err_addr``` contains the psm2 internal service id,  then the server will fail on line 1287, because that fi_getinfo() needs the user-passed-in fake service id, but src_err_addr contains the psm2 internal service id. I didn't find a way to get around that.

Reply:
Hmm ok I'll take a look. It should be fixed. Also you're inserting directly into the AV which is probably not right, it should go through the hash table stuff.

Reply:
> Also you're inserting directly into the AV which is probably not right
You are right. the address should be inserted to the hash table.  On the client side I packed the ip and port on line 2280, is there a way to retrieve that on the server side for the recv_unexpected_event case? or do we need to use a req header like the sockets provider?

Reply:
I don't think you should use the req header. Do you really need to pack the IP/port to get that to work?

Reply:
that's the only way I can think of for the client to pass its ip:port uri to the server. The server passes its URI to the client by the port.cfg file, the client can't do it that way. Do you have any suggestions?